### ISSUE TITLE
pyannote.audio 4.0.3 uses 6x more VRAM than 3.3.2 (>9.54GB vs 2.59GB peak)

### ISSUE THREAD

**Issue by DevAtShopot** [pyannote/pyannote-audio#1963]
### Tested versions

  | Version              | Model                           | VRAM Peak | Reproducible  |
  |----------------------|---------------------------------|-----------|---------------|
  | pyannote.audio 3.3.2 | speaker-diarization-3.1         | 1.59GB    | ✅ No issue    |
  | pyannote.audio 4.0.3 | speaker-diarization-community-1 | 9.54GB    | ✅ Bug present |
  | pyannote.audio 4.0.3 | speaker-diarization-3.1         | 9.54GB    | ✅ Bug present |

  The VRAM spike occurs regardless of which model is used with 4.0.3.

### System information

  - OS: Ubuntu 22.04 LTS   - GPU: NVIDIA RTX A5000 (24GB VRAM)   - Python: 3.12   - PyTorch: 2.5.1+cu124   - CUDA: 12.x

### Issue description

When running speaker diarization on a 72-minute audio file, pyannote.audio 4.0.3 uses 6x more VRAM than 3.3.2.

  Step-by-step VRAM comparison

  | Processing Step      | 3.3.2 + diarization-3.1 | 4.0.3 + community-1 |
  |----------------------|-------------------------|---------------------|
  | segmentation         | 0.40GB                  | 0.43GB              |
  | embeddings           | 0.05GB                  | 0.05GB              |
  | discrete_diarization | 1.59GB                  | 9.54GB              |

  The spike occurs during the discrete_diarization step (after clustering, during reconstruction).

  Expected behavior

  VRAM usage should be comparable between versions for similar workloads.

  Actual behavior

  4.0.3 allocates ~8GB more VRAM during reconstruction, making it impractical for GPUs with less than 12GB or for concurrent
  processing.

 Additional observations

  1. Both versions use identical:
    - Embedding model: pyannote/wespeaker-voxceleb-resnet34-LM
    - Clustering: AgglomerativeClustering
    - batch_size=32
  2. The issue persists whether audio is passed as file path or preloaded waveform
  3. In 4.0.3, exclusive_speaker_diarization is always computed even when legacy=True, potentially contributing to overhead

### Minimal reproduction example (MRE)

Minimal reproduction example (MRE)    import os   import torch   import torchaudio    torch.cuda.set_device(0)   torch.cuda.empty_cache()    from pyannote.audio import Pipeline    def log_mem(label):       peak = torch.cuda.max_memory_allocated()/1024**3       print(f"{label}: peak={peak:.2f}GB")    # Load model   pipeline = Pipeline.from_pretrained(       "pyannote/speaker-diarization-3.1",       token=os.environ.get("HF_TOKEN")  # or use_auth_token for 3.3.2   )   pipeline.to(torch.device("cuda"))    # Load audio (use any 60+ minute audio file)   audio_file = "your_audio.wav"   waveform, sample_rate = torchaudio.load(audio_file)    # Track memory at each step   last_step = [None]   def hook(step_name, step_artefact, file, completed=None, total=None):       if step_name != last_step[0]:           log_mem(f"Step: {step_name}")           last_step[0] = step_name           torch.cuda.reset_peak_memory_stats()    torch.cuda.reset_peak_memory_stats()   result = pipeline({"waveform": waveform, "sample_rate": sample_rate}, hook=hook)    print(f"\nFinal peak VRAM: {torch.cuda.max_memory_allocated()/1024**3:.2f}GB")    Results    pyannote.audio 3.3.2:   Step: segmentation: peak=0.40GB   Step: embeddings: peak=0.05GB   Step: discrete_diarization: peak=1.59GB   Final peak VRAM: 0.04GB    pyannote.audio 4.0.3:   Step: segmentation: peak=0.43GB   Step: embeddings: peak=0.05GB   Step: discrete_diarization: peak=9.54GB   Final peak VRAM: 0.04GB    ---   Impact    - Users with 8-12GB GPUs cannot run 4.0.3 on long audio files   - Concurrent processing severely limited (only ~2 jobs vs ~6+ with 3.3.2 on 24GB GPU)   - Forces users to stay on 3.3.2 (which requires pinning huggingface_hub<=0.23.5)

**Comment by mcflyspb** [comment]
I tested both diarization pipelines in the same environment and switching from `pyannote/speaker-diarization-community-1` to `pyannote/speaker-diarization-3.1` had almost no effect on GPU memory usage.

**Environment**

* OS: Ubuntu 24.04.4 LTS
* Kernel: Linux 6.8.0-90-generic
* GPUs: 2x NVIDIA RTX A5000 (24564 MiB each)
* NVIDIA driver: 580.126.09
* `nvidia-smi` CUDA version: 13.0
* `nvcc` version: 12.0.140
* Python interpreter used for the pipeline: `./env/bin/python`
* PyTorch: `2.11.0+cu130`
* torchaudio: `2.11.0+cu130`
* torchvision: `0.26.0+cu130`
* pyannote.audio: `4.0.4`
* pyannote.core: `6.0.1`
* pyannote.pipeline: `4.0.0`
* soundfile: `0.13.1`
* transformers: `4.57.6`
* accelerate: `1.12.0`
* safetensors: `0.7.0`
* huggingface_hub: `0.36.2`
* numpy: `2.4.3`
* scipy: `1.17.1`

**Configurations tested**

`DIAR_MODEL = os.getenv("DIAR_MODEL", "pyannote/speaker-diarization-community-1")`
`ASR_MODEL = os.getenv("ASR_MODEL", "Qwen/Qwen3-ASR-1.7B")`

`DIAR_MODEL = os.getenv("DIAR_MODEL", "pyannote/speaker-diarization-3.1")`
`ASR_MODEL = os.getenv("ASR_MODEL", "Qwen/Qwen3-ASR-1.7B")`

**Input**

* WAV, 16 kHz mono
* duration: `3751.0s`
* diarization chunk size: `7200s`
* device: `cuda:1`
* the whole file is processed as a single diarization chunk

**Comparison: `community-1` vs `speaker-diarization-3.1`**

**Diarization phase**

* after diarizer load `allocated`: `0.03 GiB` vs `0.03 GiB`
* after diarizer load `reserved`: `0.05 GiB` vs `0.05 GiB`
* after `waveform.to(device)` `allocated`: `0.25 GiB` vs `0.25 GiB`
* after `waveform.to(device)` `reserved`: `0.27 GiB` vs `0.27 GiB`
* `max_allocated` inside `diarizer()`: `10.53 GiB` vs `10.53 GiB`
* `max_reserved` inside `diarizer()`: `12.09 GiB` vs `12.09 GiB`
* after diarization chunk `allocated`: `0.04 GiB` vs `0.04 GiB`
* after diarization chunk `reserved`: `0.05 GiB` vs `0.05 GiB`

**ASR phase**

* after ASR load: `3.81 GiB allocated / 4.39 GiB reserved`
* both runs reached the ASR phase with effectively identical GPU memory usage

**Timing**

* `community-1`: `~59.0s`
* `speaker-diarization-3.1`: `~59.7s`

**Output difference**

* `community-1`: `357 segments`
* `speaker-diarization-3.1`: `336 segments`

The model change affects segmentation output, but does not reduce the GPU memory spike in this setup.

**Result**

Replacing `pyannote/speaker-diarization-community-1` with `pyannote/speaker-diarization-3.1` did not reduce the peak GPU memory usage.

Observed peak in both cases:

* `max_allocated = 10.53 GiB`
* `max_reserved = 12.09 GiB`

This suggests that the main VRAM spike is likely caused by a shared part of the diarization pipeline or post-processing path rather than by the specific diarization model weights.

**Comment by Karen86Tonoyan** [comment]
Thanks, this is very useful data.

Your measurements strongly suggest that the VRAM spike is not primarily
caused by the specific diarization model checkpoint, since both
`community-1` and `speaker-diarization-3.1` reach essentially the same peak
memory usage.

Given that:
- input is processed as a single 3751s chunk,
- peak happens inside `diarizer()`,
- post-run memory returns to a very low baseline,

it looks more likely that the memory spike comes from a shared
inference/post-processing path in the diarization pipeline rather than from
model weight size alone.

So the practical lever for reducing VRAM may be:
- reducing diarization chunk length,
- forcing smaller internal windows/batching if supported,
- CPU/offloaded post-processing where possible,
- or changing pipeline structure rather than only swapping the diarization
model.

wt., 7 kwi 2026, 09:48 użytkownik mcflyspb ***@***.***>
napisał:

**Comment by mcflyspb** [comment]
## GPU memory timeline for `./env/bin/python`

Below is a timeline for **only the `./env/bin/python` process (PID 4003908)**.  
Function call timestamps come from the application runtime log, while GPU memory values come from `nvidia-smi` sampled every ~2–3 seconds. As a result, memory windows should be treated as **observed**, not millisecond-exact. :contentReference[oaicite:0]{index=0}

### Peak memory

- **Peak GPU memory of the process:** **12,642 MiB**
- **Peak window:** approximately **14:10:19 → 14:10:42**
- **Phase:** `run_diarization_phase()` → internal call to `diarizer()`

This matches the internal PyTorch memory stats from stdout:

- `max_allocated = 10.53 GiB`
- `max_reserved = 12.09 GiB`

---

## Function timeline

### Startup and diarization load

- `14:08:33.502` — `main()` started, device selected: `cuda:1`
- `14:08:33.660` — `init_cuda_for_stats()` finished
- `14:08:33.660` — `load_audio_16k_mono()` started
- `14:08:34.742` — `load_diarizer()` started
- `14:08:41.320` — diarization model loaded

### Diarization phase

- `14:08:41.320` — `run_diarization_phase()` started
- `14:08:41.321` — waveform preparation for diarization chunk
- `14:08:41.474` — `waveform.to(device)` finished
- `14:08:41.474` — `diarizer()` entered
- `14:10:42.361` — `diarizer()` returned
- `14:10:42.364` — annotation parsed, `336` segments found
- `14:10:44.819` — cleanup after diarization chunk
- `14:10:44.824` — `run_diarization_phase()` finished

### Transition to ASR

- `14:10:44.825` — diarizer unload started
- `14:10:45.721` — diarizer unload finished
- `14:10:45.723` — `load_asr()` started for `distil-whisper/distil-large-v3`
- `14:11:17.518` — ASR model loaded
- `14:11:17.519` — `prepare_segments_for_asr()` started
- `14:11:17.520` — `run_asr_phase()` started
- `14:11:18.900` — first recognized segment logged

---

## Process-only GPU memory timeline

### 1. Process appears on GPU

- `14:08:35` — **206 MiB** :contentReference[oaicite:1]{index=1}

This corresponds to early process startup after CUDA initialization and before heavy diarization compute begins.

### 2. Early diarization growth

- `14:08:42` — **1,242 MiB** :contentReference[oaicite:2]{index=2}
- `14:08:47` — **2,506 MiB** :contentReference[oaicite:3]{index=3}
- `14:08:50` to `14:10:17` — stable around **2,514 MiB** :contentReference[oaicite:4]{index=4}

This is the pre-peak working set while `run_diarization_phase()` is active and before the largest allocation spike is observed.

### 3. Peak during diarization

- `14:10:19` — **12,642 MiB** :contentReference[oaicite:5]{index=5}
- `14:10:22` — **12,642 MiB** :contentReference[oaicite:6]{index=6}
- `14:10:25` — **12,642 MiB** :contentReference[oaicite:7]{index=7}
- `14:10:27` — **12,642 MiB** :contentReference[oaicite:8]{index=8}
- `14:10:30` — **12,642 MiB** :contentReference[oaicite:9]{index=9}
- `14:10:32` — **12,642 MiB** :contentReference[oaicite:10]{index=10}
- `14:10:35` — **12,642 MiB** :contentReference[oaicite:11]{index=11}
- `14:10:37` — **12,642 MiB** :contentReference[oaicite:12]{index=12}
- `14:10:40` — **12,642 MiB** :contentReference[oaicite:13]{index=13}
- `14:10:42` — **12,642 MiB** :contentReference[oaicite:14]{index=14}

This is the highest observed GPU memory usage of the process and aligns with the tail end of the `diarizer()` call.

### 4. Immediate release after diarization

- `14:10:45` — **312 MiB** :contentReference[oaicite:15]{index=15}
- `14:10:47` — **282 MiB** :contentReference[oaicite:16]{index=16}
- `14:10:50` — **282 MiB** :contentReference[oaicite:17]{index=17}
- `14:11:07` — **282 MiB** :contentReference[oaicite:18]{index=18}

This matches diarizer unload and explicit cleanup.

### 5. ASR model load and early inference

- `14:11:10` — **1,840 MiB** :contentReference[oaicite:19]{index=19}
- `14:11:17` — **1,888 MiB** :contentReference[oaicite:20]{index=20}
- `14:11:20` — **1,914 MiB** :contentReference[oaicite:21]{index=21}
- `14:11:25` — **1,914 MiB** :contentReference[oaicite:22]{index=22}
- `14:11:27` — **1,906 MiB** :contentReference[oaicite:23]{index=23}
- `14:11:30` — **1,906 MiB** :contentReference[oaicite:24]{index=24}

At the point where the available logs end, the ASR stage is running with a working set of roughly **1.84–1.91 GiB**.

---

## Condensed summary

- The process starts on GPU at **206 MiB**
- During diarization it first grows to about **2.5 GiB**
- The **maximum observed process memory** is **12.6 GiB**, during `diarizer()`
- After diarization cleanup it drops to about **282–312 MiB**
- During ASR loading and early inference it stabilizes near **1.9 GiB**

---

## Conclusion

For this run, the GPU memory bottleneck is clearly **speaker diarization**, not ASR:  
**~12.6 GiB peak during `diarizer()` vs ~1.9 GiB during early Whisper ASR**.

**Comment by mcflyspb** [comment]
The large GPU memory spike is caused by the embedding stage, not segmentation.

From the instrumented run:
- `_segmentation` peaks at only ~0.63 GiB allocated / ~0.97 GiB reserved
- repeated `_embedding` calls usually peak at ~1.82–1.83 GiB
- one specific `_embedding` call spikes to ~10.53 GiB allocated / ~12.09 GiB reserved

After that, `clustering` and `to_diarization` do not increase active GPU memory further. The high `reserved` value persists because of PyTorch's CUDA caching allocator, while `allocated` drops back to ~0.26–0.29 GiB.

**Comment by mcflyspb** [comment]
The GPU memory spike happens in the embedding stage, specifically during the final `_embedding(...)` call.

Normal embedding calls use:
- input shape `(32, 1, 160000)`
- masks shape `(32, 589)`
- peak allocated memory around `1.82–1.83 GiB`

The problematic final call uses:
- input shape `(29, 1, 160000)`
- masks shape `(29, 589)`
- peak allocated memory `10.53 GiB`
- peak reserved memory `12.09 GiB`

Since the time dimension stays the same (`160000`) and only the batch size changes (`32 -> 29`), the spike is most likely caused by a different CUDA/cuDNN algorithm/workspace selection for the final partial batch, not by a longer input sequence.

**Comment by mcflyspb** [comment]
We observed a very large GPU memory spike while running diarization with `pyannote/speaker-diarization-3.1`.

At first this looked like a memory leak, but after adding detailed instrumentation around the pipeline stages, it turned out that the spike happens specifically during the speaker embedding step (`_embedding`), not during segmentation, clustering, ASR, or waveform upload.

With the default pipeline settings, the embedding stage was executed with relatively large batches (up to 29–32 items per call), for example:

* `_embedding.arg0.shape = (29, 1, 160000)`
* `_embedding.masks.shape = (29, 589)`

During those calls, GPU usage jumped to approximately:

* `max_allocated ~= 10.53 GiB`
* `max_reserved ~= 12.09 GiB`

After the diarization chunk finished, memory dropped back close to baseline, which suggests this is not a leak but a large temporary allocation / workspace spike inside the embedding stage.

To test the hypothesis, we overrode the pipeline setting:

```python
diarizer.embedding_batch_size = 4
```

After that, the embedding calls became:

* `_embedding.arg0.shape = (4, 1, 160000)`

and the memory spike almost disappeared:

* `max_allocated ~= 0.46 GiB`
* `max_reserved ~= 0.97 GiB`

The rest of the pipeline behavior remained consistent.

So the root cause of the GPU memory spike appears to be the speaker embedding batch size used inside `speaker-diarization-3.1`.

This does not look like a leak in user code because:

* memory returns to near-baseline after the chunk is finished
* the spike is localized to `_embedding`
* reducing `embedding_batch_size` directly removes the problem

Why this matters:

On GPUs with limited VRAM, the default embedding batch size can make the pipeline unexpectedly unstable or unusable, even for workloads that otherwise fit in memory.

Suggested improvements:

1. Expose `embedding_batch_size` more explicitly in the documentation / examples
2. Use a safer default for lower-memory GPUs
3. Optionally adapt batch size automatically based on available VRAM

Environment:

* Model: `pyannote/speaker-diarization-3.1`
* GPU: CUDA device
* Input: 16 kHz mono WAV
* Diarization run in a chunked pipeline
* Peak traced with custom debug instrumentation around pipeline stages

Minimal workaround:

```python
diarizer = Pipeline.from_pretrained("pyannote/speaker-diarization-3.1", ...)
diarizer.embedding_batch_size = 4
```

This significantly reduces peak GPU memory usage.
